### DISCUSSION TITLE
Would you be interested in JavaScript components?

### DISCUSSION THREAD

**Discussion by aubymori** [zapSNH/zapsCoolPhotonTheme#13] [Ideas]
@zapSNH This is a nice theme you have here! One of the best I've seen in a while. I worked on the [Echelon](https://github.com/echelon-theme/echelon) theme, as well as my own Phroton theme (which was intended to be a successor to YukisCoffee's, but ultimately never got finished), and both of these incorporate JavaScript components to increase the extent to which Firefox can be themed.

Advantages:
- DOM modifications (like panel arrows!)
- Updating theme to match user prefs without restarting the browser
- Custom `about` protocol pages (like a configuration page)

Disadvantages:
- Most likely would not work with the extension version, as all of this is browser-level JS, with access to XPCOM.

**Comment by aubymori** [comment]
![image](https://github.com/user-attachments/assets/27057596-8eca-40c8-8673-aefdbb659fed)
Here's an old screenshot from my Phroton theme. As you can see, it has arrows on panels, and the old Photon appmenu.

**Reply by aubymori** [reply]
![image](https://github.com/user-attachments/assets/f0c19a16-83fa-4f91-93e7-0a2e69f64180)

**Comment by zapSNH** [comment]
Looks coolio and thanks!

I was aware of JS components to modify the browser (like Echelon or [duskFox](https://github.com/aminomancer/uc.css.js)) before but I didn't really bother since learning how to do this kind of stuff seemed a bit too much for me at the time. 

I'll prolly look into it once/if I have the time (i'd also appreciate some help too, but i think i could manage on my own). The theme and  extension are usually developed with Nightly to quickly keep up with changes so I'm curious on how that would mess with stuff.

I do have some questions:
1. Are you sure that it wouldn't work with the extension version? The extension is very loosely based on [Paxmod](https://github.com/numirias/paxmod), which used to do/does pretty cool stuff like multi-row tabs and tab title coloring.
2. If you don't mind me asking, why was the new Phroton theme never finished? It looks pretty cool (and more complete than mine lol).

**Reply by aubymori** [reply]
1. Actually looking at Paxmod, it seems to have access to the browser window JS scope, which *should* have the necessary access to XPCOM? I'd have to actually test further (which I don't feel like doing right now).
2. I was basically done with the browser UI itself and couldn't figure out theming the dialogs in preferences, which eventually lead to me just stopping work on the theme.

**Reply by aubymori** [reply]
Actually, nevermind. Looks like the API Mozilla provides for extensions modifying browser chrome is limited.
